File: fushi/test/pages/popup_selected_dictionary_oneshot_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// TODO-645 / BUG-358: the popup mining dictionary selection must be one-shot.
///
/// 根因：`fushi/assets/popup/popup.js` 的 `selectedDictionaries[idx]`（记某词条
/// 长按选了哪本词典，填 Anki `{selected-glossary}` 字段）原来只在用户长按同一词典
/// 取消时清。制卡成功 / 换词重渲染都不清，于是换词复用常驻热槽 WebView（同 entryIdx）
/// 时残留选择粘到下一张卡，静默带上次选的词典。修复对齐句子上下文镜像的一次性生命周期：
/// 制卡成功后清本词条（`resetSelectedDictionariesForEntry`），换词由宿主调
/// `resetSelectedDictionaries()` 清全部。
///
/// 这里两层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `buildMinePayload`（读
///    `selectedDictionaries[idx]?.name` 进 `selectedDictionary` 字段），断言
///    制卡成功清后 / 换词 reset 后该字段为空，且多词条选择相互独立。无 node 时 skip。
/// ② 源码级——静态扫描 popup.js + 宿主，保证一次性清理的两入口与宿主调用都在位
///    （即便无 node 也能在 CI 守住回归）。
void main() {
  test(
    'popup dictionary selection is one-shot (executes buildMinePayload via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/pages/popup_selected_dictionary_oneshot_test.js',
      );
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'popup dictionary one-shot JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  test('popup.js clears the dictionary selection after a successful mine', () {
    final String js = File('assets/popup/popup.js').readAsStringSync();

    // The reset helpers must exist (mirror of resetSentenceContextMirror).
    expect(
      js,
      contains('window.resetSelectedDictionaries = function()'),
      reason: 'word-change reset helper must exist',
    );
    expect(
      js,
      contains('window.resetSelectedDictionariesForEntry = function(idx)'),
      reason: 'per-entry mine-success reset helper must exist',
    );

    // The mine-success branch (after the real mineEntry call) must clear THIS
    // entry's selection. Assert the per-entry clear sits AFTER the mineEntry call
    // (so the mined card still carried the pick) and BEFORE the post-mine
    // anki refresh closure — i.e. inside the success path.
    final int mineCall = js.indexOf('const reply = await mineEntry(');
    expect(mineCall, greaterThanOrEqualTo(0),
        reason: 'mine-success call missing');
    final int perEntryClear =
        js.indexOf('window.resetSelectedDictionariesForEntry(idx);', mineCall);
    expect(
      perEntryClear,
      greaterThan(mineCall),
      reason:
          'mine-success branch must clear this entry selection after mineEntry',
    );
    final int refreshClosure = js.indexOf('const refreshFromAnki', mineCall);
    expect(
      perEntryClear,
      lessThan(refreshClosure),
      reason:
          'per-entry clear must be in the mine-success path before the refresh closure',
    );
  });

  test('host word-change inject zeros the dictionary selection', () {
    final String dart = File(
      'lib/src/pages/implementations/dictionary_popup_webview.dart',
    ).readAsStringSync();

    expect(
      dart,
      contains('window.resetSelectedDictionaries();'),
      reason: 'host must zero the dictionary selection on word change',
    );

    // It must sit right next to the existing sentence-context mirror reset and
    // before renderPopup() (the same reused-warm-slot reason).
    final int sentenceReset =
        dart.indexOf('window.resetSentenceContextMirror();');
    final int dictReset = dart.indexOf('window.resetSelectedDictionaries();');
    final int render = dart.indexOf('window.renderPopup();', sentenceReset);
    expect(sentenceReset, greaterThanOrEqualTo(0));
    expect(dictReset, greaterThan(sentenceReset));
    expect(render, greaterThan(dictReset),
        reason:
            'dictionary reset must run before renderPopup rebuilds the DOM');
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/pages/popup_selected_dictionary_oneshot_test.js
// TODO-645 / BUG-358 behavior test: the popup's per-entry dictionary selection
// (`selectedDictionaries[idx]`, which fills the Anki {selected-glossary} field)
// must be ONE-SHOT — cleared after a successful mine and on word change — so it
// never sticks to the next card mined from the same (reused warm-slot) WebView.
//
// Regression context: `selectedDictionaries[idx]` is set on a long-press of a
// dictionary header and was only ever cleared when the user long-pressed the
// SAME header again to toggle it off. A successful mine (and re-querying a new
// word, which reuses the warm popup WebView at the same entryIdx) left the stale
// selection in place, so the next mined card silently carried the previously
// chosen dictionary. This mirrors the sentence-context-mirror lifecycle, which
// is already zeroed both after mine and on word change.
//
// This test EXECUTES the real popup.js against a minimal fake DOM. It drives the
// real `buildMinePayload` (which reads `selectedDictionaries[idx]?.name` into the
// `selectedDictionary` field) using an entry with no glossaries so the DOM-heavy
// glossary/frequency/pitch builders short-circuit. Reverting the fix (leaving the
// selection sticky / removing the reset helpers) turns this red.
//
// Run: node fushi/test/pages/popup_selected_dictionary_oneshot_test.js
// (also driven from popup_selected_dictionary_oneshot_test.dart so it executes
//  inside `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const popupPath = path.resolve(__dirname, '../../assets/popup/popup.js');
const yomitanRendererPath = path.resolve(
  __dirname,
  '../../assets/popup/yomitan-glossary-renderer.js',
);
const source = fs.readFileSync(yomitanRendererPath, 'utf8') + '\n' +
  fs.readFileSync(popupPath, 'utf8');

// A throwaway DOM element good enough for the popup helpers that might touch the
// DOM. (With empty glossaries none are actually exercised, but createElement
// must hand back something with the methods popup.js calls.)
function makeElement(tag) {
  return {
    tagName: (tag || 'div').toUpperCase(),
    className: '',
    id: '',
    textContent: '',
    innerHTML: '',
    style: {},
    dataset: {},
    children: [],
    attributes: [],
    classList: {
      _set: new Set(),
      add(name) { this._set.add(name); },
      remove(name) { this._set.delete(name); },
      contains(name) { return this._set.has(name); },
    },
    appendChild(child) { this.children.push(child); return child; },
    append(...nodes) { this.children.push(...nodes); },
    setAttribute() {},
    removeAttribute() {},
    addEventListener() {},
    querySelectorAll() { return []; },
    querySelector() { return null; },
    closest() { return null; },
  };
}

function makeSandbox() {
  const documentObj = {
    documentElement: { style: {}, classList: makeElement().classList },
    head: { appendChild() {} },
    body: makeElement('body'),
    getElementById() { return null; },
    querySelector() { return null; },
    querySelectorAll() { return []; },
    createElement(tag) { return makeElement(tag); },
    addEventListener() {},
  };

  const windowObj = {
    audioSources: [],
    needsAudio: false,
    lookupEntries: [],
    dictionaryStyles: {},
    flutter_inappwebview: { callHandler() { return Promise.resolve(false); } },
    getSelection() { return { toString() { return ''; } }; },
  };
  documentObj.defaultView = windowObj;

  const sandbox = {
    Node: { TEXT_NODE: 3, ELEMENT_NODE: 1 },
    Date, Math, URL, JSON, RegExp, Set, Map, Object, Array, console,
    performance: { now() { return 0; } },
    setTimeout, clearTimeout,
    DOMParser: class { parseFromString() { return { body: makeElement('body'), querySelectorAll() { return []; } }; } },
    document: documentObj,
    window: windowObj,
    getComputedStyle() { return {}; },
  };
  sandbox.globalThis = sandbox;
  return sandbox;
}

// Load the real popup.js into a fresh sandbox and surface module-scope symbols.
function loadPopup() {
  const sandbox = makeSandbox();
  vm.createContext(sandbox);
  const exported = source + `
    ;window.__test = {
      buildMinePayload: buildMinePayload,
      getSelected: function(idx) { return selectedDictionaries[idx]; },
      selectDictionary: function(idx, name) {
        selectedDictionaries[idx] = { name: name, label: { classList: { remove: function() {} } } };
      },
      selectedDictionaryNames: function() {
        return Object.keys(selectedDictionaries).map(function(k) {
          return k + '=' + selectedDictionaries[k].name;
        });
      },
    };
  `;
  vm.runInContext(exported, sandbox, { filename: 'popup.js' });
  return sandbox;
}

function makeEntry(expression, reading) {
  return { expression: expression, reading: reading, glossaries: [], frequencies: [], pitches: [] };
}

async function payloadFor(sandbox, idx) {
  const e = makeEntry('猫', 'ねこ');
  return await sandbox.window.__test.buildMinePayload(
    e.expression, e.reading, e.frequencies, e.pitches, [], '猫', idx, '猫');
}

(async function run() {
  // Test 1: a long-pressed dictionary populates {selected-glossary} (sanity),
  // and after the per-entry mine-success clear the same idx mines empty.
  {
    const sb = loadPopup();
    sb.window.__test.selectDictionary(0, 'Daijirin');
    const before = await payloadFor(sb, 0);
    assert.strictEqual(before.selectedDictionary, 'Daijirin',
      'a long-pressed dictionary must populate {selected-glossary}; got '
        + JSON.stringify(before.selectedDictionary));

    assert.strictEqual(typeof sb.window.resetSelectedDictionaries, 'function',
      'popup.js must expose window.resetSelectedDictionaries (mirror of '
        + 'resetSentenceContextMirror) for the word-change reset path');
    assert.strictEqual(typeof sb.window.resetSelectedDictionariesForEntry, 'function',
      'popup.js must expose window.resetSelectedDictionariesForEntry for the '
        + 'per-entry mine-success clear path');

    sb.window.resetSelectedDictionariesForEntry(0);
    const after = await payloadFor(sb, 0);
    assert.strictEqual(after.selectedDictionary, '',
      'after a successful mine the SAME entry must mine an empty '
        + 'selectedDictionary (one-shot); got ' + JSON.stringify(after.selectedDictionary));
  }

  // Test 2: word change → resetSelectedDictionaries zeros ALL entries.
  {
    const sb = loadPopup();
    sb.window.__test.selectDictionary(0, 'Daijirin');
    sb.window.__test.selectDictionary(1, 'JMdict');
    assert.deepStrictEqual(
      sb.window.__test.selectedDictionaryNames().sort(),
      ['0=Daijirin', '1=JMdict'],
      'precondition: two entries selected');

    sb.window.resetSelectedDictionaries();

    assert.deepStrictEqual(sb.window.__test.selectedDictionaryNames(), [],
      'resetSelectedDictionaries must clear EVERY entry selection on word change');

    const p0 = await payloadFor(sb, 0);
    const p1 = await payloadFor(sb, 1);
    assert.strictEqual(p0.selectedDictionary, '',
      'after word-change reset, idx0 mine must carry empty selectedDictionary; got '
        + JSON.stringify(p0.selectedDictionary));
    assert.strictEqual(p1.selectedDictionary, '',
      'after word-change reset, idx1 mine must carry empty selectedDictionary; got '
        + JSON.stringify(p1.selectedDictionary));
  }

  // Test 3: per-idx independence — clearing one entry must not touch the other.
  {
    const sb = loadPopup();
    sb.window.__test.selectDictionary(0, 'Daijirin');
    sb.window.__test.selectDictionary(1, 'JMdict');

    sb.window.resetSelectedDictionariesForEntry(0);

    const p0 = await payloadFor(sb, 0);
    const p1 = await payloadFor(sb, 1);
    assert.strictEqual(p0.selectedDictionary, '',
      'mining idx0 must clear idx0 selection; got ' + JSON.stringify(p0.selectedDictionary));
    assert.strictEqual(p1.selectedDictionary, 'JMdict',
      'mining idx0 must NOT clear idx1 selection (per-entry independence); got '
        + JSON.stringify(p1.selectedDictionary));
  }

  console.log('popup_selected_dictionary_oneshot_test.js: all assertions passed');
})().catch(function(e) {
  console.error(e && e.stack ? e.stack : e);
  process.exit(1);
});
